core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -325,24 +325,26 @@ final case class NonEmptyList[+A](head: A, tail: List[A]) {
     *
     * {{{
     * scala> import cats.data.NonEmptyList
+    * scala> import cats.instances.boolean._
     * scala> val nel = NonEmptyList.of(12, -2, 3, -5)
     * scala> nel.groupBy(_ >= 0)
     * res0: Map[Boolean, cats.data.NonEmptyList[Int]] = Map(false -> NonEmptyList(-2, -5), true -> NonEmptyList(12, 3))
     * }}}
     */
-  def groupBy[B](f: A => B): Map[B, NonEmptyList[A]] = {
-    val m = mutable.Map.empty[B, mutable.Builder[A, List[A]]]
+  def groupBy[B](f: A => B)(implicit B: Order[B]): Map[B, NonEmptyList[A]] = {
+    var m = TreeMap.empty[B, mutable.Builder[A, List[A]]](B.toOrdering)
+
     for { elem <- toList } {
-      m.getOrElseUpdate(f(elem), List.newBuilder[A]) += elem
-    }
-    val b = immutable.Map.newBuilder[B, NonEmptyList[A]]
-    for { (k, v) <- m } {
-      val head :: tail = v.result // we only create non empty list inside of the map `m`
-      b += ((k, NonEmptyList(head, tail)))
+      val k = f(elem)
+
+      m.get(k) match {
+        case None => m += ((k, List.newBuilder[A] += elem))
+        case Some(builder) => builder += elem
+      }
     }
-    b.result
-  }
 
+    m.mapValues(v => NonEmptyList.fromListUnsafe(v.result))
```

Comment:
IIRC, `mapValues` is by-name and full of surprises and not something you would ever want to use.

Reply:
The alternative would be to construct another `TreeMap`. Calling `map` instead of `mapValues` wouldn't work because we'd have to call `toMap` on it at the end, and that uses the universal equality and hashing we're trying to get rid off.

What do you think?

Reply:
BTW, I never had problems with `mapValues`, which is why I used it.

Reply:
Another surprise of `.mapValues` is that the result is not Serializable.

Reply:
Why don't you use `Functor[Map].map(m)(v => NonEmptyList.fromListUnsafe(v.result))`? :)

Reply:
Erm, of course meant:

```scala
Functor[Map[B, ?]].map(m)(v => NonEmptyList.fromListUnsafe(v.result))
```

Reply:
Ok, I think was wrong when I said we can't use `map` instead of `mapValues`. The result of `TreeMap.map` is a `TreeMap` as well, which means the ordering is preserved. I'll change to that.

I don't think we need to redundant `Functor[Map[B, ?]]` summoning here.